File: app/token_bucket.lua
-- KEYS[1] = bucket key (e.g. "bucket:client123")
-- ARGV[1] = max_tokens
-- ARGV[2] = refill_rate (tokens per second)
-- ARGV[3] = current timestamp (seconds, float)

local key = KEYS[1]
local max_tokens = tonumber(ARGV[1])
local refill_rate = tonumber(ARGV[2])
local now = tonumber(ARGV[3])

local bucket = redis.call("HMGET", key, "tokens", "last_refill")
local tokens = tonumber(bucket[1])
local last_refill = tonumber(bucket[2])

if tokens == nil then
    -- first request from this client: bucket starts full
    tokens = max_tokens
    last_refill = now
end

-- refill based on elapsed time
local elapsed = now - last_refill
local refill_amount = elapsed * refill_rate
tokens = math.min(max_tokens, tokens + refill_amount)

local allowed = 0
if tokens >= 1 then
    tokens = tokens - 1
    allowed = 1
end

redis.call("HMSET", key, "tokens", tokens, "last_refill", now)
redis.call("EXPIRE", key, 3600)  -- clean up inactive buckets after 1hr

return allowed

File: app/token_bucket.py
# app/token_bucket.py
import time


class TokenBucket:
    def __init__(self, max_tokens: int, refill_rate: float):
        self.max_tokens = max_tokens
        self.refill_rate = refill_rate
        self.tokens = max_tokens
        self.last_refill_time = time.monotonic()

    def _refill(self):
        now = time.monotonic()
        elapsed = now - self.last_refill_time
        refill_amount = elapsed * self.refill_rate
        self.tokens = min(self.max_tokens, self.tokens + refill_amount)
        self.last_refill_time = now

    def allow_request(self) -> bool:
        self._refill()
        if self.tokens >= 1:
            self.tokens -= 1
            return True
        return False